### PR TITLE
Issue #2143: Check that import groups aren't separated internally

### PR DESCRIPTION
**Pull request by linelect**
Issue: [#2143](https://github.com/checkstyle/checkstyle/issues/2143)

Add new check, and it give a lot of "Redundant separation" errors, fixed.

### CODE REVIEW

**Comment by romani** [comment]
Good, but please split functional changes of Check and fixes in other files in different commits. It is ok in this case to have two commits in one PR.

**Comment by MEZk** [comment]
@romani 

Why did you decide to implement new check instead of fixing CustomImportOrder (https://github.com/checkstyle/checkstyle/issues/3551) and ImportOrder (https://github.com/checkstyle/checkstyle/issues/2143)?

**Comment by MEZk** [comment]
@linelect 
Don't forget to update integration tests and documentation.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 1]
```diff
@@ -195,6 +195,12 @@
      */
     public static final String MSG_ORDERING = "import.ordering";
 
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_REDUNDANT_SEPARATION = "import.groups.separated.internally";
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: redundant vs internally ...... you already do not know what you do.
Check should enforce no separation in group so it should be "separation in group".
MSG_REDUNDANT_SEPARATION -> MSG_SEPARATED_IN_GROUP
import.groups.separated.internally - > import.groups.separated.in.group

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 2]
```diff
@@ -443,12 +449,26 @@ else if (groupIdx > lastGroup) {
         else {
             log(line, MSG_ORDERING, name);
         }
-
+        checkRedundantSeparation(groupIdx, isStatic, line, name);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: should be checkSeparatorInGroup

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 3]
```diff
@@ -443,12 +449,26 @@ else if (groupIdx > lastGroup) {
...
+     * Checks whether group of imports have redundant separation
+     * @param groupIdx group number.
+     * @param isStatic whether the token is static or not.
+     * @param line the line of the current import.
+     * @param name the name of the current import.
+     */
+    private void checkRedundantSeparation(int groupIdx, boolean isStatic, int line, String name) {
+        if (!beforeFirstImport && separated && groupIdx == lastGroup
+                && isStatic == lastImportStatic && line - lastImportLine > 1) {
+            log(line, MSG_REDUNDANT_SEPARATION, name);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: please try to return boolean from this variable and do "log" in visitToken to let easily see when log of violation happen.

File: src/main/resources/com/puppycrawl/tools/checkstyle/checks/imports/messages.properties [hunk 4]
```diff
@@ -10,6 +10,7 @@ import.separation=''{0}'' should be separated from previous imports.
 import.control.missing.file=Missing an import control file.
 import.control.disallowed=Disallowed import - {0}.
 import.control.unknown.pkg=Import control file does not handle this package.
+import.groups.separated.internally=Redundant separation in import group before ''{0}''
```
Comment (romani) [inline src/main/resources/com/puppycrawl/tools/checkstyle/checks/imports/messages.properties]: Redundant --> Extra.
@rnveach , please suggest us better wording.

File: src/xdocs/config_imports.xml [hunk 5]
```diff
@@ -1159,6 +1159,10 @@ public class InputEclipseStaticImportsOrder { }
       <subsection name="Error Messages">
         <ul>
           <li>
+            <a href="https://github.com/search?q=path%3Asrc%2Fmain%2Fresources%2Fcom%2Fpuppycrawl%2Ftools%2Fcheckstyle%2Fchecks%2Fimports+filename%3Amessages*.properties+repo%3Acheckstyle%2Fcheckstyle+%22import.groups.separated.internally%22">
+            import.groups.separated.internally</a>
+          </li>
+          <li>
```
Comment (romani) [inline src/xdocs/config_imports.xml]: please update xdoc description with new functionality, please add example (users like examples more than descriptions).

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 6]
```diff
@@ -195,6 +195,12 @@
      */
     public static final String MSG_ORDERING = "import.ordering";
 
+    /**
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: please update javadoc description of this Check with new functionality.

**Comment by romani** [comment]
@linelect , please fix referenced items and reply "done" or ".....any reason...." for EACH of them

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 1 above)
Comment (linelect) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 2 above)
Comment (linelect) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 3 above)
Comment (linelect) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: done

File: src/main/resources/com/puppycrawl/tools/checkstyle/checks/imports/messages.properties (hunk 4 above)
Comment (linelect) [inline src/main/resources/com/puppycrawl/tools/checkstyle/checks/imports/messages.properties]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 6 above)
Comment (linelect) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: done

File: src/xdocs/config_imports.xml (hunk 5 above)
Comment (linelect) [inline src/xdocs/config_imports.xml]: I update config_imports.xml, but i don't assured that it's all that need. For new functionality don't need add additional options to examples config.

**Comment by romani** [comment]
@rnveach , please do review.

**Pull request merged**